web-api/src/storage/elastic.rs

```diff
@@ -339,3 +357,160 @@ struct IngestedDocument<'a> {
+) -> KnnSearchParts {
+    let inner_filter =
+        create_es_search_filter(params.time, params.excluded, params.published_after);
+
+    let knn_object = create_knn_request_object(
+        params.embedding,
+        params.count,
+        params.num_candidates,
+        &inner_filter,
+    );
+
+    let mut generic_parameters = json_object!({
+        "size": params.count
+    });
+
+    if let Some(min_score) = params.min_similarity {
+        generic_parameters.extend(json_object!({
+            "min_score": min_score,
+        }));
+    }
+
+    KnnSearchParts {
+        knn_object,
+        generic_parameters,
+        inner_filter,
+    }
+}
+
+fn create_es_search_filter<'a>(
+    reference_time: DateTime<Utc>,
+    excluded_ids: impl IntoIterator<Item = &'a DocumentId>,
+    published_after: Option<DateTime<Utc>>,
+) -> JsonObject {
+    let time = reference_time.to_rfc3339();
+
+    // the existing documents are not filtered in the query to avoid too much work for a cold
+    // path, filtering them afterwards can occasionally lead to less than k results though
+    let excluded_ids = json!({
+        "ids": {
+            "values": excluded_ids.into_iter().collect_vec()
+        }
+    });
+
+    if let Some(published_after) = published_after {
+        // published_after != null && published_after <= publication_date <= time
+        json_object!({
+            "filter": {
+                "range": {
+                    "properties.publication_date": {
+                        "gte": published_after.to_rfc3339(),
+                        "lte": time
+                    }
+                }
+            },
+            "must_not": excluded_ids
+        })
+    } else {
+        // published_after == null || published_after <= time
+        json_object!({
+            "must_not": [
+                excluded_ids,
+                {
+                    "range": {
+                        "properties.publication_date": {
+                            "gt": time
+                        }
+                    }
+                }
+            ]
+        })
+    }
+}
+
+fn create_knn_request_object(
+    embedding: &NormalizedEmbedding,
+    count: usize,
+    num_candidates: usize,
+    filter: &JsonObject,
+) -> JsonObject {
+    // https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html
+    json_object!({
+        "knn": {
+            "field": "embedding",
+            "query_vector": embedding,
+            "k": count,
+            "num_candidates": num_candidates,
+            "filter": {
+                "bool": filter
+            }
+        }
+    })
+}
+
+fn normalize_scores<K>(mut scores: HashMap<K, f32>) -> HashMap<K, f32>
+where
+    K: Eq + Hash,
+{
+    let max_score = scores
+        .values()
+        .max_by(|l, r| l.total_cmp(r))
+        .copied()
+        .unwrap_or_default();
+    // TODO[pmk/now] the original code did a .max(1.0); this looked like bug, verify
```

Comment:
For some reason the max score clause for bm25 normalization had a `max(1.0)`, this seems strange as I would thing especially with values >1 we would want to normalize bm25.

Reply:
I for some reason did read `value.max(1)` as "value maximally 1". :facepalm:  It's kinda an interesting case as when having `std::cmp::max` in mind you would never mistake it but also reading it a "value maximally 1" doesn't seem wrong at all either.

> in light of the other discussions/question though, we probably should apply max-normalization uniformly for all 

probably, or use rrf